File: Java/TP3_MET_JAVA/src/Iteratif/MotVideIteratif.java
package Iteratif;

import Interface.Mot;

public enum MotVideIteratif implements Mot {

    SINGLETON;

    public boolean casVide() {
        return true;
    }

    public Character caractere() {
        throw new UnsupportedOperationException();
    }

    public Mot reste() {
        throw new UnsupportedOperationException();
    }

    @Override
    public Mot creerMotVide() {
        return MotVideIteratif.SINGLETON;
    }

    @Override
    public Mot creerCaractereMot(char c) {
        return new CaractereMotIteratif(c, this);
    }

    @Override
    public Mot creerMotMot(Mot droite) {
        return new MotMotIteratif(this, droite);
    }
}

File: TypeScript/TP3_MET_TS/src/Iteratif/MotVideIteratif.ts

import {CaractereMotIteratif} from "./CaractereMotIteratif";
import {MotMotIteratif} from "./MotMotIteratif";
import {Mot} from "../Interface/Mot";

export class MotVideIteratif extends Mot {

    public static SINGLETON: MotVideIteratif = new MotVideIteratif();

    public casVide(): boolean {
        return true;
    }

    public iterator(): Iterator<string> {
        return this.iterateurIteratif();
    }

    public creerMotVide(): Mot {
        return MotVideIteratif.SINGLETON;
    }

    public creerCaractereMot(c: string): Mot {
        return new CaractereMotIteratif(c, this);
    }

    public creerMotMot(droite: Mot): Mot {
        return new MotMotIteratif(this, droite);
    }
}
